src/rest_framework_reactive/observer.py

```diff
@@ -135,53 +135,46 @@ def evaluate(self, return_full=True, return_emitted=False):
         try:
             settings = get_queryobserver_settings()
 
-            for retry in range(MAX_INTEGRITY_ERROR_RETRIES):
+            with transaction.atomic():
+                observer_qs = models.Observer.objects.select_for_update()
                 try:
+                    # We need another transaction to keep to outer one intact
+                    # if create fails.
                     with transaction.atomic():
-                        # Obtain the observer state and lock it. This prevents an observer from being
-                        # processed in parallel from multiple different workers.
-                        observer, _ = models.Observer.objects.select_for_update().get_or_create(
-                            id=self.id,
-                            defaults={'request': pickle.dumps(self._request)},
+                        observer = observer_qs.create(
+                            id=self.id, request=pickle.dumps(self._request)
                         )
-
-                        # Evaluate the observer.
-                        start = time.time()
-                        result = self._evaluate(observer, return_full, return_emitted)
-                        duration = time.time() - start
-
-                        # Log slow observers.
-                        if duration > settings['warnings']['max_processing_time']:
-                            logger.warning(
-                                "Slow observed viewset ({})".format(
-                                    self._get_logging_id()
-                                ),
-                                extra=self._get_logging_extra(duration=duration),
-                            )
-
-                        # Stop really slow observers.
-                        if duration > settings['errors']['max_processing_time']:
-                            logger.error(
-                                "Stopped extremely slow observed viewset ({})".format(
-                                    self._get_logging_id()
-                                ),
-                                extra=self._get_logging_extra(
-                                    stopped=True, duration=duration
-                                ),
-                            )
-                            observer.delete()
-
-                        return result
                 except IntegrityError:
-                    # If an IntegrityError occurrs we need to rollback the transaction and retry observer
-                    # evaluation as another transaction may be creating an observer concurrently.
-                    if retry == MAX_INTEGRITY_ERROR_RETRIES - 1:
-                        raise
+                    observer = observer_qs.get(id=self.id)
```

Comment:
Can observer be deleted between create and select? Should the atomic block encapsulate try/catch?

Reply:
When I made this comment I only notices the atomic block in line 144, and not the one in 138.